Correct the location of the recalculation of the FIS directory size,
and also add the same recalculation for the byte-swapped case.

Signed-off-by: Rod Whitby <rod@whitby.id.au>

---
 drivers/mtd/redboot.c |    8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

Index: linux-2.6.20-arm/drivers/mtd/redboot.c
===================================================================
--- linux-2.6.20-arm.orig/drivers/mtd/redboot.c
+++ linux-2.6.20-arm/drivers/mtd/redboot.c
@@ -96,6 +96,8 @@
 			 */
 			if (swab32(buf[i].size) == master->erasesize) {
 				int j;
+				/* Update numslots based on actual FIS directory size */
+				numslots = swab32(buf[i].size) / sizeof (struct fis_image_desc);
 				for (j = 0; j < numslots; ++j) {
 
 					/* A single 0xff denotes a deleted entry.
@@ -120,11 +122,11 @@
 					swab32s(&buf[j].desc_cksum);
 					swab32s(&buf[j].file_cksum);
 				}
+			} else {
+				/* Update numslots based on actual FIS directory size */
+				numslots = buf[i].size / sizeof(struct fis_image_desc);
 			}
 			break;
-		} else {
-			/* re-calculate of real numslots */
-			numslots = buf[i].size / sizeof(struct fis_image_desc);
 		}
 	}
 	if (i == numslots) {
